src/lib/operators/table_scan/sorted_segment_between_search.hpp

```diff
@@ -0,0 +1,193 @@
+#pragma once
+
+#include <functional>
+#include <thread>
+#include <type_traits>
+
+#include <boost/range.hpp>
+#include <boost/range/join.hpp>
+
+#include "all_type_variant.hpp"
+#include "constant_mappings.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+// Generic class which handles the actual scanning of a sorted segment
+template <typename IteratorType, typename SearchValueType>
+class SortedSegmentBetweenSearch {
+ public:
+  SortedSegmentBetweenSearch(IteratorType begin, IteratorType end, const OrderByMode& order_by, const bool nullable,
+                             const PredicateCondition& predicate_condition, const SearchValueType& left_value,
+                             const SearchValueType& right_value)
+      : _begin{begin},
+        _end{end},
+        _predicate_condition{predicate_condition},
+        _left_value{left_value},
+        _right_value{right_value},
+        _nullable{nullable},
+        _is_ascending{order_by == OrderByMode::Ascending || order_by == OrderByMode::AscendingNullsLast},
+        _is_nulls_first{order_by == OrderByMode::Ascending || order_by == OrderByMode::Descending} {}
+
+ private:
+  void _exponential_search_for_nulls(IteratorType it_first, IteratorType it_last) {
+    if (it_first == it_last) return;
+    // If no null values are present
+    if (_is_nulls_first && !it_first->is_null()) {
+      return;
+    }
+    if (!_is_nulls_first && !(it_last - 1)->is_null()) {
+      return;
+    }
+
+    using difference_type = typename std::iterator_traits<IteratorType>::difference_type;
+    difference_type size = std::distance(it_first, it_last), bound = 1;
+
+    if (_is_nulls_first) {
+      while (bound < size && (it_first + bound)->is_null()) {
+        bound *= 2;
+      }
+
+      auto end = it_first + std::min(bound, size);
+      _begin = std::lower_bound(it_first + (bound / 2), end, false,
+                                [](const auto& segment_position, const auto& _) { return segment_position.is_null(); });
```

Comment:
(I believe here we should use `std::upper_bound - 1` as well, see comment to line 60..61 in this same file)

Reply:
(Update from https://github.com/hyrise/hyrise/pull/1961#discussion_r367366937 applies here as well)